source/val/validate_layout.cpp

```diff
@@ -302,7 +302,7 @@ spv_result_t FunctionScopedInstructions(ValidationState_t& _,
   } else {
     return _.diag(SPV_ERROR_INVALID_LAYOUT, inst)
            << spvOpcodeString(opcode)
-           << " cannot appear in a function declaration";
+           << " is in an invalid module layout section";
```

Comment:
This function is checking the function declarations specifically.  If the instruction that is invalid is in this position then it is in function declaration.  Why make the message more general by say it is in an invalid layout section?

This will make it even harder for developers to pick out which is the incorrect instruction and why.

Reply:
Because this doesn't just run for things inside a function. I spent a bunch of time trying to figure out the below error as I _thought_ it had something to do with functions but it doesn't.

```
~/Development/tint [main]$ ../spirv-tools/out/Debug/tools/spirv-val out/shader.bad.spv 
error: line 7: EntryPoint cannot appear in a function declaration
  OpEntryPoint Fragment %fragmentUint "fragmentUint" %colorUint
```

vs

```
~/Development/tint [main]$ ../spirv-tools/out/Debug/tools/spirv-val out/shader.bad.spv
error: line 7: EntryPoint is in an invalid module layout section
  OpEntryPoint Fragment %fragmentUint "fragmentUint" %colorUint
```

The problem shader section was:

```
               OpEntryPoint Fragment %fragmentFloat "fragmentFloat" %colorFloat
               OpExecutionMode %fragmentFloat OriginUpperLeft
               OpEntryPoint Fragment %fragmentUint "fragmentUint" %colorUint
               OpExecutionMode %fragmentUint OriginUpperLeft
```

Reply:
Ah, maybe this isn't the best way to fix this. Reading through the code I think what is happening is because the order of sections goes backwards, the |ModuleScopedInstructions| method moves us all the way to the |FunctionDeclaration| scope when we shouldn't be. So, we end up in a weird spot that we have non-function scoped declarations but we think we're in the function declaration.

Reply:
Validation of layout sections can only progress forwards. If you place an instruction from an earlier section after an instruction from a later section, the validator always ends up at function layout and issuing this error. That leads to confusion for instructions in the module scope sections.

Ideally the validator could do a better job reporting this error. Something like the following, but less ugly in `ModuleScopedInstructions` before the while loop:
```
for (ModuleLayoutSection section = kLayoutCapabilities; 
       section != _.current_layout_section(); ++section) {
  if (IsInstructionInLayoutSection(section, opcode)) { // note this is a private helper currently
    // New function for stringifying the section.
    return _.Diag(SPV_ERROR_INVALID_LAYOUT, inst) 
      << "Instruction can only appear in " << getSectionName(section) 
      << " section, but occurs after " << getSectionName(_.current_layout_section()) 
      << " section instructions";
  }
}
```